src/arfsdao.ts

```diff
@@ -152,57 +154,115 @@ export class ArFSDAOAnonymous extends ArFSDAOType {
-			const txData = await this.arweave.transactions.getData(driveBuilder.txId, { decode: true });
-			const dataString = await Utf8ArrayToStr(txData);
-			const dataJSON = await JSON.parse(dataString);
+		const { data } = response.data;
+		const { transactions } = data;
+		const { edges } = transactions;
 
-			// Get the drive name and root folder id
-			driveBuilder.name = dataJSON.name;
-			driveBuilder.rootFolderId = dataJSON.rootFolderId;
+		if (!edges.length) {
+			throw new Error(`Public folder with Folder ID ${folderId} not found!`);
 		}
 
-		return driveBuilder.build();
+		const folderBuilder = new ArFSPublicFolderBuilder(folderId, this.arweave);
+		return await folderBuilder.build();
+	}
+
+	async getPublicFilesWithParentFolderIds(folderIDs: FolderID[]): Promise<ArFSPublicFile[]> {
+		let cursor = '';
+		let hasNextPage = true;
+		const allFiles: ArFSPublicFile[] = [];
+
+		while (hasNextPage) {
+			const gqlQuery = buildQuery(
+				[
+					{ name: 'Parent-Folder-Id', value: folderIDs },
+					{ name: 'Entity-Type', value: 'file' }
+				],
+				cursor
+			);
+
+			const response = await this.arweave.api.post(graphQLURL, gqlQuery);
+			const { data } = response.data;
+			const { transactions } = data;
+			const { edges } = transactions;
+			hasNextPage = transactions.pageInfo.hasNextPage;
+			const files: Promise<ArFSPublicFile>[] = edges.map(async (edge: GQLEdgeInterface) => {
+				const { node } = edge;
+				cursor = edge.cursor;
+				const fileBuilder = await ArFSPublicFileBuilder.fromArweaveNode(node, this.arweave);
+				return fileBuilder.build(node);
+			});
+			allFiles.push(...(await Promise.all(files)));
+		}
+		return allFiles;
+	}
+
+	async getAllFoldersOfPublicDrive(driveId: DriveID): Promise<ArFSPublicFolder[]> {
+		let cursor = '';
+		let hasNextPage = true;
+		const allFolders: ArFSPublicFolder[] = [];
+
+		while (hasNextPage) {
+			const gqlQuery = buildQuery(
+				[
+					{ name: 'Drive-Id', value: driveId },
+					{ name: 'Entity-Type', value: 'folder' }
+				],
+				cursor
+			);
+
+			const response = await this.arweave.api.post(graphQLURL, gqlQuery);
+			const { data } = response.data;
+			const { transactions } = data;
+			const { edges } = transactions;
+			hasNextPage = transactions.pageInfo.hasNextPage;
+			const folders: Promise<ArFSPublicFolder>[] = edges.map(async (edge: GQLEdgeInterface) => {
+				const { node } = edge;
+				cursor = edge.cursor;
+				const folderBuilder = ArFSPublicFolderBuilder.fromArweaveNode(node, this.arweave);
+				return await folderBuilder.build(node);
+			});
+			allFolders.push(...(await Promise.all(folders)));
+		}
+		return allFolders;
+	}
+
+	/**
+	 * Lists the children and self of certain public folder
+	 * @param {FolderID} folderId the folder ID to list children of
+	 * @returns {ArFSPublicFileOrFolderWithPaths[]} an array representation of the children and parent folder
+	 */
+	async listPublicFolder(folderId: FolderID): Promise<ArFSPublicFileOrFolderWithPaths[]> {
```

Comment:
Why is this function:
a) returning the parent folder?
b) returning an entire subtree?

The manpage entry under the CRD looks like this:
```
list-folder:
  • Gets all ArFS Metadata and Data Txids for the files and folders within this folder
  --folder-id: the ArFS folder ID for the folder to query
  --drive-password: the drive password for drive of the folder identified by --drive-id
  • Required only for folders residing in private drives
  • Can NOT be used in conjunction with --drive-key
--drive-key: the drive key for the parent drive of the folder identified by --folder-id
  • Required only for folders residing in private drives
  • Can NOT be used in conjunction with --drive-password
  --get-all-revisions: (OPTIONAL) gets every revision of each ArFS entity within this folder
  --all: (OPTIONAL) gets all contents within this folder, including child files/folders.
  --max-depth: (OPTIONAL) enter a number of the amount of sub folder levels to list
  --verbose: (OPTIONAL) returns all metadata for the returned entities
```

`list-folder` as I interpreted it was intended to only list the immediate children of the folder. The default --max-depth, therefore, would be 1 (or 0 depending on how we want to do the indexing. We could check common linux commands for inspo). --all would be used to get the entire contents of the folder hierarchy below the parent folder. Your implementation seems to be `--all` by default, which is a much more expensive operation for users who might only want a depth of 1. And in either of those cases, I don't understand why the parent folder would be returned. I support one argument might be that `ls -al` returns `.` and `..` and that `find .` returns `.` as well. I'm open to suggestions.

I think an efficient implementation here could:
• Check the depth. If it's 1, use a simpler GQL query, i.e. all entities with parentFolderId set to the provided one and you're done.
OTHERWISE:
• If the depth is any deeper than 1, get the folder hierarchy for the drive (as you've done here) and get the list of folders at a particular depth (you'd need a new function on the FolderHierarchy class for this).
• Do a GQL request for all files with parent folderIDs from the trimmed folder list and return those files and folders.

I suppose we can take what you have for now and add more stories for the --max-depth and --all flag separation.